src/librustc_typeck/check/mod.rs

```diff
@@ -1763,47 +1759,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
             debug!("select_all_obligations_and_apply_defaults: defaults={:?}", default_map);
 
-            // We loop over the unsolved variables, resolving them and if they are
-            // and unconstrainted numberic type we add them to the set of unbound
-            // variables. We do this so we only apply literal fallback to type
-            // variables without defaults.
+            // Examine all unsolved variables, and narrow them to the set that have applicable
+            // defaults. We want to process any unsolved variables that have either an explicit
+            // user default, literal fallback, or are diverging.
             for ty in &unsolved_variables {
                 let resolved = self.infcx().resolve_type_vars_if_possible(ty);
-                if self.infcx().type_var_diverges(resolved) {
-                    demand::eqtype(self, codemap::DUMMY_SP, *ty, self.tcx().mk_nil());
-                } else {
-                    match self.infcx().type_is_unconstrained_numeric(resolved) {
-                        UnconstrainedInt | UnconstrainedFloat => {
-                            unbound_tyvars.insert(resolved);
-                        },
-                        Neither => {}
-                    }
-                }
-            }
-
-            // We now remove any numeric types that also have defaults, and instead insert
-            // the type variable with a defined fallback.
-            for ty in &unsolved_variables {
                 if let Some(_default) = default_map.get(ty) {
-                    let resolved = self.infcx().resolve_type_vars_if_possible(ty);
 
                     debug!("select_all_obligations_and_apply_defaults: ty: {:?} with default: {:?}",
                              ty, _default);
 
                     match resolved.sty {
-                        ty::TyInfer(ty::TyVar(_)) => {
-                            unbound_tyvars.insert(ty);
-                        }
-
-                        ty::TyInfer(ty::IntVar(_)) | ty::TyInfer(ty::FloatVar(_)) => {
-                            unbound_tyvars.insert(ty);
-                            if unbound_tyvars.contains(resolved) {
-                                unbound_tyvars.remove(resolved);
-                            }
+                        ty::TyInfer(_) => {
+                            unbound_tyvars.insert(*ty);
```

Comment:
This set `unbound_tyvars` probably needs a better name, like `defaults_to_apply`